lib/load.php

```diff
@@ -24,31 +24,36 @@
 require dirname( __FILE__ ) . '/register.php';
 require dirname( __FILE__ ) . '/demo.php';
 
-// Register server-side code for individual blocks.
-if ( ! function_exists( 'render_block_core_archives' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/archives/index.php';
-}
-if ( ! function_exists( 'render_block_core_block' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/block/index.php';
-}
-if ( ! function_exists( 'render_block_core_categories' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/categories/index.php';
-}
-// Currently merged in core as `gutenberg_render_block_core_latest_comments`,
-// expected to change soon.
-if ( ! function_exists( 'render_block_core_latest_comments' )
-	&& ! function_exists( 'gutenberg_render_block_core_latest_comments' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/latest-comments/index.php';
-}
-if ( ! function_exists( 'render_block_core_latest_posts' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/latest-posts/index.php';
-}
-if ( ! function_exists( 'render_block_core_rss' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/rss/index.php';
-}
-if ( ! function_exists( 'render_block_core_shortcode' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/shortcode/index.php';
-}
-if ( ! function_exists( 'render_block_core_search' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/search/index.php';
+/**
+ * Discovers block files from the plugin built artifact, unregistering the
+ * equivalent in core if already defined, then includes the block file.
+ */
+function gutenberg_reregister_core_block_types() {
+	// Blocks directory may not exist if working from a fresh clone.
+	$blocks_dir = dirname( __FILE__ ) . '/../build/block-library/blocks/';
+	if ( ! file_exists( $blocks_dir ) ) {
+		return;
+	}
+
+	$registry = WP_Block_Type_Registry::get_instance();
+
+	$files = scandir( $blocks_dir );
+	foreach ( $files as $file ) {
+		$parts = pathinfo( $file );
+
+		// Ignore all non-PHP files, subdirectories, path traversal.
+		if ( 'php' !== $parts['extension'] ) {
+			continue;
+		}
+
+		// Derive the assumed block name by file path.
+		$block_name = 'core/' . $parts['filename'];
+
+		if ( $registry->is_registered( $block_name ) ) {
+			$registry->unregister( $block_name );
+		}
+
+		require $blocks_dir . $file;
```

Comment:
So this function is called on the `init` hook and the files it requires also trigger functions on the `init` hook. Isn't it too late for the `init` hooks added in those files to trigger properly?

Reply:
> So this function is called on the `init` hook and the files it requires also trigger functions on the `init` hook. Isn't it too late for the `init` hooks added in those files to trigger properly?

Mmm, I'd have sworn I'd seen somewhere we'd used a similar implementation, but in mind of the following, I think it'd be expected to _not_ work. If we handle `init` here at an earlier priority, it may not be an issue though.

- https://github.com/WordPress/wordpress-develop/blob/83275aa/tests/phpunit/tests/actions.php#L350
- https://core.trac.wordpress.org/ticket/17817
- https://core.trac.wordpress.org/ticket/38011#comment:2

Reply:
I figured out that my comment is (https://github.com/WordPress/gutenberg/pull/13521#issuecomment-464031003) related to what @youknowriad raised. When I change the code to execute the block's registration immediately then all blocks load properly.

Reply:
> Isn't it too late for the init hooks added in those files to trigger properly?

That should work ever since `WP_Hook` was introduced.

Reply:
> That should work ever since `WP_Hook` was introduced.

Can you clarify what you mean @swissspidy ? The current issue is that we want to load the blocks implementations at the same priority as the core registration, which based on the [aforementioned test cases](https://github.com/WordPress/wordpress-develop/blob/83275aa/tests/phpunit/tests/actions.php#L350) does not appear to be a supported use-case for hooks.